File: 150-evaluate-reverse-polish-notation/150-evaluate-reverse-polish-notation.scala
import scala.collection.mutable._

object Solution {
    def evalRPN(tokens: Array[String]): Int = {
        val stack: Stack[Int] = Stack()
        
        tokens.foreach { tok => 
            tok match {
                case "+" => stack.push(stack.pop + stack.pop) 
                case "-" => {
                    val a = stack.pop
                    val b = stack.pop
                    stack.push(b - a)
                }
                case "*" => stack.push(stack.pop * stack.pop) 
                case "/" => {
                    val a = stack.pop
                    val b = stack.pop
                    stack.push(b / a)
                }
                case n => stack.push(n.toInt)
            }
        }
        
        stack.top
    }
}

File: 150-evaluate-reverse-polish-notation/150-evaluate-reverse-polish-notation.java
class Solution {
    public int evalRPN(String[] tokens) {
        Deque<Integer> stack = new LinkedList<>();
        
        for(int i = 0; i < tokens.length; i++) {
            String tok = tokens[i];
            
            if (tok.equals("+") || tok.equals("-") || tok.equals("*") || tok.equals("/")) {
                int v2 = stack.pop();
                int v1 = stack.pop();
                
                if (tok.equals("+")) stack.push(v1 + v2);
                else if (tok.equals("-")) stack.push(v1 - v2);
                else if (tok.equals("*")) stack.push(v1 * v2);
                else stack.push(v1 / v2);
                
                continue;
            }
            
            stack.push(Integer.valueOf(tok));
        }
        
        return stack.peek();
    }
}